### PR TITLE
Update nomicon for variance typo & contravariance

### LINKED ISSUE #28100: Update Rustonomicon to mention contravariance
[Claiming that rust does not use contravariance in its type system](https://doc.rust-lang.org/nightly/nomicon/subtyping.html#variance) seems to be an error.

The Rustonomicon should be updated. I've already been confused enough since I took the "it does not exist" as a fact.

Quoting irc:
> `< eddyb> playbot: fn foo<'a, T: 'static>(x: &'a fn(&'a T), ptr: &mut &'a fn(&'static T)) { *ptr = x; }`
>  `-playbot:#rust- ()`
>  `< eddyb> if that's not contravariance, I don't know what is`

### LINKED ISSUE #28704: Possible error in example on subtyping and variance in nomicon
The following piece of code is present at the end of variance chapter in nomicon:

```rust
use std::cell::Cell;

struct Foo<'a, 'b, A: 'a, B: 'b, C, D, E, F, G, H> {
    a: &'a A,     // variant over 'a and A
    b: &'b mut B, // invariant over 'b and B
    c: *const C,  // variant over C
    d: *mut D,    // invariant over D
    e: Vec<E>,    // variant over E
    f: Cell<F>,   // invariant over F
    g: G,         // variant over G
    h1: H,        // would also be variant over H except...
    h2: Cell<H>,  // invariant over H, because invariance wins
}
```

It says that `Foo` is invariant over `'b`, however, all references, mutable included, are variant over their lifetime parameters, as explained earlier in this chapter. Wouldn't then `Foo` be variant over `'b`?

### PR DESCRIPTION
**Pull request by bluss**
Update nomicon for variance typo & contravariance

Fixes #28704
Fixes #28100

### CODE REVIEW

**Comment by bluss** [comment]
This updates nomicon to just mention contravariance without having a longer explanation.

I will cast my own doubts on contravariance for fn() to be removed, since it will mean some completely valid trait implementations will stop compiling.

It's a simple change for a parenthesis sentence, but I'll ping the language team anyway (sorry!)

cc @rust-lang/lang

File: src/doc/nomicon/subtyping.md [hunk 1]
```diff
@@ -44,10 +44,11 @@ subtyping of its outputs. There are two kinds of variance in Rust:
...
 (For those of you who are familiar with variance from other languages, what we
-refer to as "just" variance is in fact *covariance*. Rust does not have
-contravariance. Historically Rust did have some contravariance but it was
-scrapped due to poor interactions with other features. If you experience
-contravariance in Rust call your local compiler developer for medical advice.)
+refer to as "just" variance is in fact *covariance*. Rust has *contravariance*
+for functions. The future of contravariance is uncertain and it may be
+scrapped. For now, `fn(T)` is contravariant in `T`, which is used in matching
+methods in trait implementations to the trait definition. Traits are
+not covariant in their parameters, so `Fn(T)`, is invariant in `T`).
```
Comment (Gankro) [inline src/doc/nomicon/subtyping.md]: not contravariant?

File: src/doc/nomicon/subtyping.md [hunk 2]
```diff
@@ -44,10 +44,11 @@ subtyping of its outputs. There are two kinds of variance in Rust:
 * F is *invariant* over `T` otherwise (no subtyping relation can be derived)
 
 (For those of you who are familiar with variance from other languages, what we
-refer to as "just" variance is in fact *covariance*. Rust does not have
-contravariance. Historically Rust did have some contravariance but it was
-scrapped due to poor interactions with other features. If you experience
-contravariance in Rust call your local compiler developer for medical advice.)
+refer to as "just" variance is in fact *covariance*. Rust has *contravariance*
+for functions. The future of contravariance is uncertain and it may be
+scrapped. For now, `fn(T)` is contravariant in `T`, which is used in matching
```
Comment (Gankro) [inline src/doc/nomicon/subtyping.md]: I feel like this can't just be aside anymore

File: src/doc/nomicon/subtyping.md (hunk 1 above)
Comment (bluss) [inline src/doc/nomicon/subtyping.md]: thanks, fixed.

File: src/doc/nomicon/subtyping.md (hunk 2 above)
Comment (bluss) [inline src/doc/nomicon/subtyping.md]: IMO depends on if it's getting scrapped or not.

**Comment by Gankro** [comment]
@bors r+ rollup

Needs work, but a clear improvement.

File: src/doc/nomicon/subtyping.md [hunk 3]
```diff
@@ -44,10 +44,11 @@ subtyping of its outputs. There are two kinds of variance in Rust:
...
+methods in trait implementations to the trait definition. Traits don't have
+inferred variance, so `Fn(T)`, is invariant in `T`).
```
Comment (durka) [inline src/doc/nomicon/subtyping.md]: > `Fn(T)`,

extra comma

**Comment by bluss** [comment]
Fixed the comma, and let's see if bors can test this

@bors r=gankro rollup

File: src/doc/nomicon/subtyping.md (hunk 3 above)
Comment (bluss) [inline src/doc/nomicon/subtyping.md]: Thanks for reading!

**Pull request merged**